### PR TITLE
Merge reduction implementations

### PR DESCRIPTION
**Pull request by niboshi**
Merges the logic of `ReductionKernel` and `simple_reduction_function`.

### CODE REVIEW

File: cupy/core/reduction.pxi [hunk 1]
```diff
@@ -175,24 +175,94 @@ cpdef list _get_inout_args(
...
-    params = _get_kernel_params(params, args_info)
     return _get_simple_reduction_kernel(
         name, block_size, reduce_type, params, identity,
-        routine[0], routine[1], routine[2],
+        map_expr, reduce_expr, post_map_expr,
         type_preamble, input_expr, output_expr, _preamble, options)
 
 
-class simple_reduction_function(object):
+class _AbstractReductionKernel:
```
Comment (leofang) [inline cupy/core/reduction.pxi]: Is it possible to make the base class a `cdef` one?

Comment (niboshi) [reply cupy/core/reduction.pxi]: Thank you for suggestion.
Done in 5339ade1718679e8864f9803b298abbed1eee280.

**Comment by niboshi** [comment]
PTAL.

**Comment by asi1024** [comment]
Some parts of this PR are duplicate to #2697. I will review this PR after #2697 is merged.

File: cupy/core/reduction.pxi [hunk 2]
```diff
@@ -230,52 +297,58 @@ class simple_reduction_function(object):
...
+        if reduce_type is None:
+            reduce_type = _get_typename(out_types[0])
 
-        kern = _get_simple_reduction_function(
-            routine, self._params, args_info,
-            arr.dtype.type, ret.dtype.type, out_types,
+        if out_args:
+            out_type = out_args[0].dtype.type
+        else:
+            out_type = get_dtype(out_types[0]).type
```
Comment (okuta) [inline cupy/core/reduction.pxi]:
```suggestion
            out_type = out_types[0]
```

**Comment by okuta** [comment]
This is good PR.
It is OK to merge before # 2697.

File: cupy/core/reduction.pxi [hunk 3]
```diff
@@ -175,24 +175,97 @@ cpdef list _get_inout_args(
...
+
+        kern = self._get_kernel(
+            params, args_info, types,
+            map_expr, reduce_expr, post_map_expr, reduce_type,
+            block_size)
+        kern.linear_launch(
+            out_block_num * block_size, inout_args, 0, block_size, stream)
+        return ret
+
+    cpdef _get_expressions_and_types(self, in_args, out_args, dtype):
```
Comment (asi1024) [inline cupy/core/reduction.pxi]: Why not support them as `cdef` methods and write argument types explicitly?

Comment (niboshi) [reply cupy/core/reduction.pxi]: `cdef` methods cannot be overridden.
`dtype` argument does not have a specific type.

Comment (niboshi) [reply cupy/core/reduction.pxi]: Do you mean other arguments? 
I doubt typing tuples improves performance. I'll check.

Comment (niboshi) [reply cupy/core/reduction.pxi]: I only observe difference of 1-2 ns. I think it's negligible.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by niboshi** [comment]
Force-pushed a fix to remove an incorrect typing to `axis` argument.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by niboshi** [comment]
Conflict resolved.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by niboshi** [comment]
Fixed test failure.
PTAL.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**